pyiron/atomistics/structure/analyse.py

```diff
@@ -0,0 +1,72 @@
+# coding: utf-8
+# Copyright (c) Max-Planck-Institut für Eisenforschung GmbH - Computational Materials Design (CM) Department
+# Distributed under the terms of "New BSD License", see the LICENSE file.
+
+import numpy as np
+from pyiron_base import Settings
+from sklearn.cluster import AgglomerativeClustering
+
+__author__ = "Joerg Neugebauer, Sam Waseda"
+__copyright__ = (
+    "Copyright 2020, Max-Planck-Institut für Eisenforschung GmbH - "
+    "Computational Materials Design (CM) Department"
+)
+__version__ = "1.0"
+__maintainer__ = "Sam Waseda"
+__email__ = "waseda@mpie.de"
+__status__ = "production"
+__date__ = "Sep 1, 2017"
+
+s = Settings()
+
+class Analyse:
+    """ Class to analyse atom structure.  """
+    def __init__(self, ref_structure):
+        """
+        Args:
+            ref_structure (pyiron.atomistics.structure.atoms.Atoms): reference Atom structure
+        """
+        self._ref_structure = ref_structure
+
+    def __repr__(self):
+        """
+        Returns: __repr__
+        """
+        return ''
+
+    def get_layers(self, distance_threshold=0.01, id_list=None):
+        """
+        Get an array of layer numbers.
+
+        Args:
+            distance_threshold (float): Distance below which two points are
+                considered to belong to the same layer. For detailed
+                description: sklearn.cluster.AgglomerativeClustering
+            id_list (list/numpy.ndarray): List of atoms for which the layers
+                should be considered.
+
+        Returns: Array of layer numbers (same shape as structure.positions)
+
+        Example I - how to get the number of layers in each direction:
+
+        >>> structure = Project('.').create_structure('Fe', 'bcc', 2.83).repeat(5)
+        >>> print('Numbers of layers:', np.max(structure.analyse.get_layers(), axis=0)+1)
+
+        Example II - get layers of only one species:
+
+        >>> print('Iron layers:', structure.analyse.get_layers(
+        ...     id_list=structure.select_index('Fe'))
+        ... )
+        """
+        if id_list is None:
+            id_list = np.arange(len(self._ref_structure))
+        layers = []
+        for x in self._ref_structure.positions[np.array(id_list)].T:
+            cluster = AgglomerativeClustering(
+                linkage='complete',
+                n_clusters=None,
+                distance_threshold=distance_threshold
+            ).fit(x.reshape(-1,1))
+            args = np.unique(cluster.labels_, return_index=True)[1]
+            layers.append(x[args].argsort().argsort()[cluster.labels_])
```

Comment:
Nit: Once we're argsorting our argsorts, some sort of technical explanation or decomposition down to well-named methods is probably helpful 😉

Reply:
I just tried out a few more possibilities, but I don't really think I can make it easier to understand... Let me think about it a bit more.

Reply:
I'm not sure the code itself needs to be any easier, just explained in english or (better (in my opinion)) wrapped in a well-named function, either in part or whole.

Reply:
My code somewhere has a similar `argsort(argsort)` call and the following comment attached that makes it reasonably clear (to me)

```
    # argsort(argsort) looks weird, but what we're trying to do here is
    # permuting the plane indices such that label 0 corresponds to the plane
    # closest to the origin, 1 to the second closest, etc.  For that we first
    # calculate the mean x position of all labels and argsort it.  This gives a
    # mapping from the sorted indices to the labels returned by the clustering,
    # i.e. argsort(x)[0] maps the first sorted index 0 to the cluster index
    # closest to the origin, etc.  However we actually want to have the inverse
    # mapping, from cluster indices to sorted indices!  Curiously argsort also
    # inverts index permutations.  Inverting a index permutation means building
    # an array where the ith item is the index at which i appears in the
    # original permutation, which is just sorting indices into the original
    # permutation array according to the value in the permuation array. i.e.
    # argsorting.
```

feel free to adapt it.

Reply:
How fast is the agglomerative clustering btw?  I use this exact code for my own plane finding, but with DBSCAN and can't remember big problems.  Just curious, I see no need to change it if this is tested and fast, tough.  Everyone has their own pet clustering algorithm it seems. :')

Reply:
> My code somewhere has a similar `argsort(argsort)` call and the following comment attached that makes it reasonably clear (to me)
> 
> ```
>     # argsort(argsort) looks weird, but what we're trying to do here is
>     # permuting the plane indices such that label 0 corresponds to the plane
>     # closest to the origin, 1 to the second closest, etc.  For that we first
>     # calculate the mean x position of all labels and argsort it.  This gives a
>     # mapping from the sorted indices to the labels returned by the clustering,
>     # i.e. argsort(x)[0] maps the first sorted index 0 to the cluster index
>     # closest to the origin, etc.  However we actually want to have the inverse
>     # mapping, from cluster indices to sorted indices!  Curiously argsort also
>     # inverts index permutations.  Inverting a index permutation means building
>     # an array where the ith item is the index at which i appears in the
>     # original permutation, which is just sorting indices into the original
>     # permutation array according to the value in the permuation array. i.e.
>     # argsorting.
> ```
> 
> feel free to adapt it.

I copy&pasted it (even though I find it very long).

Reply:
> How fast is the agglomerative clustering btw? I use this exact code for my own plane finding, but with DBSCAN and can't remember big problems. Just curious, I see no need to change it if this is tested and fast, tough. Everyone has their own pet clustering algorithm it seems. :')

I don't really think there'd be a speed issue. There are simply two reasons why I use the agglomerative clustering: 1, I know what it does behind the scenes; 2, it doesn't have unidentified points.

Reply:
Btw @pmrv, whenever you have a code like this, I would strongly encourage you to implement it in pyiron as well. I guess we all have to get used to implementing our own tools in pyiron, but also rejecting some ideas if we think it shouldn't be in pyiron.

Reply:
> > How fast is the agglomerative clustering btw? I use this exact code for my own plane finding, but with DBSCAN and can't remember big problems. Just curious, I see no need to change it if this is tested and fast, tough. Everyone has their own pet clustering algorithm it seems. :')
> 
> I don't really think there'd be a speed issue. There are simply two reasons why I use the agglomerative clustering: 1, I know what it does behind the scenes; 2, it doesn't have unidentified points.

I agree, well-understood code is most important.

> Btw @pmrv, whenever you have a code like this, I would strongly encourage you to implement it in pyiron as well. 

Guilty as charged, it's been in my backlog.

Reply:
Sorry for leaving comments incessantly, but I changed my mind and separated the commands in more lines+the variable names are now probably easier to understand. Sorry @pmrv for creating this fuss, but in the end I found your comments too long...